### PR TITLE
Feat/hng 213 component media file upload

### LINKED ISSUE #213: [FEAT] - Components - MediaFileUpload
## Description

Create a reusable Media file upload component with drag-and-drop functionality using TypeScript (TSX) and Shadcn in a Next.js application. The component should support file selection through drag-and-drop and manual file input, displaying previews of the uploaded files with an option to delete uploaded images. It should also include a label prop and a customizable button component.

## Acceptance Criteria

- The component should support drag-and-drop file uploads.
- The component should allow manual file selection via a file input.
- Uploaded files should be displayed with previews (images).
- Each uploaded file should have a delete button to remove the file.
- The component should handle multiple file uploads.
- The component should pass the uploaded files to the parent component through a callback function.
- The component should pass the deleted file to the parent component through a callback function.
- The component should include a label prop for the file input.
- The component should accept a button component for customization.

## Requirements

- Use TypeScript (TSX) for type safety and better developer experience.
- Ensure the component is reusable and can be integrated into different parts of the application.
- Ensure accessibility for file input and drag-and-drop area.
- Include necessary and acceptable props:
  - `onUpload`: (files: File[]) => void
  - `onDelete`: (file: File) => void
  - `label`: string
  - `buttonComponent`: React.FC

## Tests

- Ensure the drag-and-drop functionality works correctly.
- Ensure manual file input selection works.
- Ensure file previews are displayed correctly after upload.
- Ensure the delete button removes the correct file and triggers the `onDelete` callback.
- Test the callback functions to ensure uploaded and deleted files are passed to the parent component.
- Validate that multiple files can be uploaded and displayed.
- Validate that the label prop is displayed correctly.
- Validate that the custom button component is rendered and functional.

## Tasks

- [ ] Create a `MediaUpload` component:
  - [ ] Implement drag-and-drop area.
  - [ ] Implement file input for manual selection.
  - [ ] Display previews of uploaded files.
  - [ ] Implement a delete button to remove uploaded files.
  - [ ] Pass uploaded files to the parent component through a callback.
  - [ ] Pass deleted files to the parent component through a callback.
  - [ ] Include a label prop for the file input.
  - [ ] Accept a custom button component.
- [ ] Create a test page to demonstrate the `MediaUpload` component.
- [ ] Write tests to validate functionality and requirements.
- [ ] Document the component usage and integration.

## LINKS
[Figma Link](https://www.figma.com/design/VEItfX6St5NSAqqNHImcxD/HNG-Boilerplate-Designs?node-id=2961-3069&t=UpTFCK1vivQxq9AE-4)

## IMAGES
Frame 1321315435
Frame 1321315435 (1)

### PR DESCRIPTION
**Pull request by Peliah**
# Description

Create a reusable Media file upload component with drag-and-drop functionality using TypeScript (TSX) and Shadcn in a Next.js application. The component should support file selection through drag-and-drop and manual file input, displaying previews of the uploaded files with an option to delete uploaded images. It should also include a label prop and a customizable button component

**Closes #213**

# Changes proposed

## What were you told to do?
. Create a component with a drag and drop functionality to insert images
. Include a button for manual file selection
. The component should handle multiple file uploads
. Pass the files (uploaded/deleted) files through the parent component
. Add a customizable button component as prop
. Also include a label prop

## What did you do?

created the MediaFileUpload component in the src/component/layout directory 
Created  function to upload a file,
Created the function to delete a file
created where i passed these values as props
I included the label as prop
I ran the command `pnpm dlx shadcn-ui@latest add button` to download the shadcn button located the the /src/component/ui directory
I added the index.test.tsx file for tests in my directory

# Check List (Check all the applicable boxes)

🚨Please review the [contribution guideline](CONTRIBUTING.md) for this repository.

- [X] My code follows the code style of this project.
- [X] This PR does not contain plagiarized content.
- [X] The title and description of the PR is clear and explains the approach.
- [X] I am making a pull request against the **dev branch** (left side).
- [X] My commit messages styles matches our requested structure.
- [X] My code additions will fail neither code linting checks nor unit test.
- [X] I am only making changes to files I was requested to.

# Screenshots/Videos

Screenshot from 2024-07-20 00-46-09
[Screencast from 07-20-2024 12:45:01 AM.webm](https://github.com/user-attachments/assets/b509544a-841c-444b-814d-50534bf7a240)

[Figma Link](https://www.figma.com/design/VEItfX6St5NSAqqNHImcxD/HNG-Boilerplate-Designs?node-id=2961-3069&t=UpTFCK1vivQxq9AE-4)

In the parent component where the MediaUpload Component is called i wrote sample to show how to call my component
`      <MediaUpload onFilesAdded={handleFilesAdded}
        onFileDeleted={handleFileDeleted}
        label="Accepts images, videos or 3D models"
        CustomButton={CustomUploadButton}
        />
        `
I also included a sample of the props passed into my component
`
const handleFilesAdded = (files: File[]) => {
    setUploadedFiles((prevFiles) => [...prevFiles, ...files]);
  };

  const handleFileDeleted = (file: File) => {
    setUploadedFiles((prevFiles) => prevFiles.filter((f) => f !== file));
  };
`

And finally my component can and must be used in a client component 
`"use client"`

image

The result of pnpm lint

image

the result of pnpm test

### CODE REVIEW

**Review by skeby — CHANGES_REQUESTED** [review]
Remove all comments
Move your component out of the layouts subfolder
Move test component to the tests folder

**Comment by Peliah** [comment]
> Remove all comments Move your component out of the layouts subfolder Move test component to the tests folder

Hi, @skeby I've fixed these problems and awaiting approval

File: src/test/index.test.tsx
Comment (skeby) [inline src/test/index.test.tsx:1]: Please name the test file properly (MediaUpload.test.tsx)
Tag me when you've done this

Comment (skeby) [reply src/test/index.test.tsx:1]: Attach screenshots of your tests and lint

Comment (Peliah) [reply src/test/index.test.tsx:1]: @skeby Changes requested have been done

**Comment by Peliah** [comment]
@skeby all requested changes have been made 
Please review

**Comment by Peliah** [comment]
@JulianaSau please what other changes are required? I've made the previous changes that was requested by @skeby

**Comment by skeby** [comment]
> @JulianaSau please what other changes are required? I've made the previous changes that was requested by @skeby

I never review your update, you comot your `changes requested` label

**Review by skeby — APPROVED** [review]

**Comment by skeby** [comment]
@PrudentBird can you merge this when Abdul gives you access again

**Review by mrcoded — APPROVED** [review]

**Pull request merged**
